tests/scripts/all.sh

```diff
@@ -692,179 +687,173 @@ make CC=gcc CFLAGS='-Werror -Wall -Wextra'
+#### 4. Build systems
+################################################################
 
-    # ARM Compiler 6 - Target ARMv7-M
-    armc6_build_test "--target=arm-arm-none-eabi -march=armv7-m"
+msg "build: Unix make, -Os (gcc)" # ~ 20s
+cleanup
+make CC=gcc CFLAGS='-Werror -Os'
 
-    # ARM Compiler 6 - Target ARMv8-A - AArch32
-    armc6_build_test "--target=arm-arm-none-eabi -march=armv8.2-a"
 
-    # ARM Compiler 6 - Target ARMv8-M
-    armc6_build_test "--target=arm-arm-none-eabi -march=armv8-m.main"
-
-    # ARM Compiler 6 - Target ARMv8-A - AArch64
-    armc6_build_test "--target=aarch64-arm-none-eabi -march=armv8.2-a"
+if uname -a | grep -F Linux >/dev/null; then
+    msg "build+test: make shared" # ~ 30s
+    cleanup
+    make SHARED=1 all check
 fi
 
-msg "build: allow SHA1 in certificates by default"
-cleanup
-cp "$CONFIG_H" "$CONFIG_BAK"
-scripts/config.pl set MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES
-make CFLAGS='-Werror -Wall -Wextra'
-msg "test: allow SHA1 in certificates by default"
-make test
-if_build_succeeded tests/ssl-opt.sh -f SHA-1
 
-msg "build: Default + MBEDTLS_RSA_NO_CRT (ASan build)" # ~ 6 min
+msg "build: cmake 'out-of-source' build" # ~ 20s
 cleanup
-cp "$CONFIG_H" "$CONFIG_BAK"
-scripts/config.pl set MBEDTLS_RSA_NO_CRT
-CC=gcc cmake -D CMAKE_BUILD_TYPE:String=Asan .
+MBEDTLS_ROOT_DIR="$PWD"
+mkdir "$OUT_OF_SOURCE_DIR"
+cd "$OUT_OF_SOURCE_DIR"
+cmake "$MBEDTLS_ROOT_DIR"
 make
 
-msg "test: MBEDTLS_RSA_NO_CRT - main suites (inc. selftests) (ASan build)"
+msg "test: cmake 'out-of-source' build" # < 10s
 make test
+cd "$MBEDTLS_ROOT_DIR"
+rm -rf "$OUT_OF_SOURCE_DIR"
 
-msg "build: Windows cross build - mingw64, make (Link Library)" # ~ 30s
-cleanup
-make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS='-Werror -Wall -Wextra' WINDOWS_BUILD=1 lib programs
 
-# note Make tests only builds the tests, but doesn't run them
-make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS='-Werror' WINDOWS_BUILD=1 tests
-make WINDOWS_BUILD=1 clean
+if [ $RUN_ARMCC -ne 0 ] && [ $YOTTA -ne 0 ]; then
+    # Note - use of yotta is deprecated, and yotta also requires armcc to be on the
+    # path, and uses whatever version of armcc it finds there.
+    msg "build: create and build yotta module" # ~ 6 min 30s
+    cleanup
+    record_status tests/scripts/yotta-build.sh
+fi
 
-msg "build: Windows cross build - mingw64, make (DLL)" # ~ 30s
-make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS='-Werror -Wall -Wextra' WINDOWS_BUILD=1 SHARED=1 lib programs
-make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS='-Werror -Wall -Wextra' WINDOWS_BUILD=1 SHARED=1 tests
-make WINDOWS_BUILD=1 clean
+
+
+################################################################
+#### 5. Slow outside Linux-x86_64: MemSan or memcheck
+################################################################
```

Comment:
Some things are missing here. If you think they're scope creep for this PR, please file an enhancement request on Github.

* The memcheck and memsan builds should run in the full config minus memory management features. We care about uninitialized memory and memory leaks in features that aren't in the default build.
* Valgrind (which we run on buildbot) has spotted memory leaks that Msan didn't spot. So I think we should run both.

Reply:
Regarding the first point, I fully agree, so I'll add it to the PR.

Regarding the second, I assume you mean Asan? (Msan only does uninitilised reads, Asan does bounds checks and (on Linux) enables LeakSanitizer.) Do you have an example? I wasn't aware this could happen, I'm surprised to learn that. Regardless, I'd rather leave that to a later stage, for example when we discuss which tests we want in the PR vs nightly/release campaigns.

Reply:
https://github.com/ARMmbed/mbedtls/pull/1545 passed `all.sh` but the [`make memcheck` build found a leak](http://buildbot.polarssl.org/builders/any%20-%20debian-wheezy-i386%20%28cmake%20full%29/builds/994/steps/memcheck/logs/stdio)